torch/_inductor/comms.py

```diff
@@ -0,0 +1,262 @@
+from collections import defaultdict
+from typing import List
+
+import torch.fx as fx
+from . import ir, scheduler
+from .analysis import get_runtime_snode
+from .dependencies import WeakDep
+from .utils import is_local, print2
+
+
+# Used to ensure that iterating over a set is deterministic
+def tuple_sorted(x):
+    return sorted(tuple(x), key=lambda x: x.name)
+
+
+def sink_waits(result: List[fx.Node]) -> List[fx.Node]:
+    """
+    Greedily moves waits as late as possible (i.e. until we reach an use). Optimal in terms of
+    communication overlap.
+    """
+    new_result = []
+    cur_waits = set()
+    for node in result:
+        if isinstance(node.meta["fusion_meta"].snode.node, ir.Wait):
+            cur_waits.add(node)
+        else:
+            for wait in tuple_sorted(cur_waits):
+                if node in wait.users:
+                    new_result.append(wait)
+                    cur_waits.remove(wait)
+
+            new_result.append(node)
+    for node in tuple_sorted(cur_waits):
+        new_result.append(node)
+    return new_result
+
+
+def raise_comms(result: List[fx.Node]) -> List[fx.Node]:
+    """
+    Greedily moves comms as early as possible (i.e. until we reach an input).
+    Optimal in terms of communication overlap.
+    """
+    new_result = []
+    cur_comms = []
+    for node in reversed(result):
+        if isinstance(node.meta["fusion_meta"].snode.node, ir.CollectiveKernel):
+            cur_comms.append(node)
+        else:
+            while len(cur_comms) > 0 and any([node in comm.args for comm in cur_comms]):
+                comm = cur_comms.pop(0)
+                new_result.append(comm)
+            new_result.append(node)
+    assert len(cur_comms) <= 1
+    for node in tuple_sorted(cur_comms):
+        new_result.append(node)
+    result = new_result[::-1]
+    return result
+
+
+def get_ancestors(node):
+    ancestors = set()
+    cur_nodes = [node]
+    while len(cur_nodes) > 0:
+        new_nodes = []
+        for node in cur_nodes:
+            for inp in node.args:
+                if inp not in ancestors:
+                    ancestors.add(inp)
+                    new_nodes.append(inp)
+        cur_nodes = new_nodes
+    return ancestors
+
+
+def get_descendants(node):
+    descendants = set()
+    cur_nodes = [node]
+    while len(cur_nodes) > 0:
+        new_nodes = []
+        for node in cur_nodes:
+            for inp in node.users:
+                if inp not in descendants:
+                    descendants.add(inp)
+                    new_nodes.append(inp)
+        cur_nodes = new_nodes
+    return descendants
+
+
+def decide_global_ordering_comms(nodes: List["scheduler.BaseSchedulerNode"]):
+    """
+    Just enforces the ordering that's in the input graph.
```

Comment:
are we able to assert today that we maintain a stable ordering from trace-time to inductor time (through all our various passes)?

Reply:
No, today, there is a separate inductor reordering pass that changes the ordering of the collectives. But anyways, we'll want to enforce our own ordering of the collectives with a pass here anyways.